### DISCUSSION TITLE
How to get all org/space roles and their assigned users?

### DISCUSSION THREAD

**Discussion by markusbalsam** [SAP/terraform-provider-cloudfoundry#180] [Q&A]
Since the [cloudfoundry_org_role](https://registry.terraform.io/providers/SAP/cloudfoundry/latest/docs/resources/org_role) and [cloudfoundry_space_role](https://registry.terraform.io/providers/SAP/cloudfoundry/latest/docs/resources/space_role) resources aren't idempotent (i.e., assigning a role to a user that already has this role results in an error) I'd like to find out which users are already assigned to certain org or space roles.

While the provider has a [cloudfoundry_role](https://registry.terraform.io/providers/SAP/cloudfoundry/latest/docs/data-sources/role) data source, it's not very useful since it requires to know the `guid` of a role, and as far as I can see there's no data source (yet) that provides a list of all roles for an org or space with their guids.

Is there any other way to achieve this using the current version (1.0.0-rc1)?

If not, are there any plans to add `cloudfoundry_[org|space]_roles` data sources to retrieve all roles for a given org or space? The V3 API has an endpoint that provides exactly this information: https://v3-apidocs.cloudfoundry.org/version/3.174.0/index.html#list-roles

**Comment by Dray56** [comment]
Hi @markusbalsam,

There is a [cloudfoundry_users](https://registry.terraform.io/providers/SAP/cloudfoundry/latest/docs/data-sources/users) data source that fetches all the users who have a valid role for that org or space. Kindly check if this can achieve the desired use case.

**Reply by markusbalsam** [reply]
Thanks @Dray56, For what I want to achieve, it is unfortunately not sufficient to only know that a user has _some_ role in an org or space, but I'd like to know the _exact_ roles a user has in the org or space.

I currently don't see a way to make a link from `cloudfoundry_users` to `cloudfoundry_role`. The latter requires the role's guid which is impossible to find out without a `cloudfoundry_roles` data source - unless, of course, I use a different tool to find it out beforehand, but that's not an option for my use case.
